File: backend/app/api/baseline_sets.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.baseline_set import BaselineLinkUpdate, BaselineSetCreate, BaselineSetResponse, CaptureAllCreate
from app.services import baseline_set as svc

router = APIRouter(prefix="/baseline-sets", tags=["baseline-sets"])


@router.get("/", response_model=list[BaselineSetResponse])
async def list_baseline_sets(project_id: uuid.UUID, db: AsyncSession = Depends(get_db)) -> list:
    return await svc.list_baseline_sets(db, project_id)


@router.post("/", response_model=BaselineSetResponse, status_code=201)
async def create_baseline_set(data: BaselineSetCreate, db: AsyncSession = Depends(get_db)):
    """A bare, empty BaselineSet — the frontend links already-existing
    standalone module baselines into it afterward, one /link call per
    module (BaselineComparison.tsx's own "Link Existing" flow)."""
    return await svc.create_baseline_set(db, data)


@router.post("/capture-all", response_model=BaselineSetResponse, status_code=201)
async def capture_all(data: CaptureAllCreate, db: AsyncSession = Depends(get_db)):
    """Captures a new Risk/Cost/ICD/Schedule baseline all at once, all tagged
    to a new named BaselineSet."""
    return await svc.capture_all(db, data)


@router.post("/link", status_code=204)
async def link_baseline(data: BaselineLinkUpdate, db: AsyncSession = Depends(get_db)) -> Response:
    """Attaches (baseline_set_id set) or detaches (baseline_set_id null) one
    already-existing standalone module baseline into a BaselineSet."""
    await svc.link_baseline(db, data)
    return Response(status_code=204)

File: frontend/src/lib/baselineSets.ts
import { useEffect, useState } from 'react'
import type { BaselineSet } from '@/modules/dashboard/types'
import { api } from './api'

export type BaselineModule = 'risk' | 'cost' | 'icd' | 'schedule'

// Controls Dashboard Phase 1b — same create/list shape as
// useScheduleSubprojects, plus the two capture actions (one-click
// "capture everything now", and manually linking an already-existing
// standalone module baseline into a set).
export function useBaselineSets(projectId: string | undefined) {
  const [baselineSets, setBaselineSets] = useState<BaselineSet[]>([])
  const [loading, setLoading] = useState(true)

  const load = async () => {
    if (!projectId) return
    setLoading(true)
    try {
      const { data } = await api.get<BaselineSet[]>('/api/v1/baseline-sets/', { params: { project_id: projectId } })
      setBaselineSets(data)
    } finally {
      setLoading(false)
    }
  }

  useEffect(() => {
    load()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [projectId])

  const captureAll = async (name: string, baselineDate: string, periodId: string, schedulePeriodId: string) => {
    const { data } = await api.post<BaselineSet>('/api/v1/baseline-sets/capture-all', {
      project_id: projectId, name, baseline_date: baselineDate, period_id: periodId, schedule_period_id: schedulePeriodId,
    })
    await load()
    return data
  }

  // A bare, empty set — used by the "Link Existing" flow (BaselineComparison.tsx)
  // to bundle already-existing standalone module baselines together, as
  // opposed to captureAll's "snapshot everything right now" path.
  const createSet = async (name: string, baselineDate: string) => {
    const { data } = await api.post<BaselineSet>('/api/v1/baseline-sets/', {
      project_id: projectId, name, baseline_date: baselineDate,
    })
    await load()
    return data
  }

  const linkBaseline = async (module: BaselineModule, baselineId: string, baselineSetId: string | null) => {
    await api.post('/api/v1/baseline-sets/link', { module, baseline_id: baselineId, baseline_set_id: baselineSetId })
  }

  return { baselineSets, loading, captureAll, createSet, linkBaseline, refetch: load }
}
